parsers/dataField/parser.go

```diff
@@ -0,0 +1,204 @@
+	}, nil
+}
+
+// Parse will parse the provided data field
+func (odp *operationDataFieldParser) Parse(dataField []byte, sender, receiver []byte) *ResponseParseData {
+	return odp.parse(dataField, sender, receiver, false)
+}
+
+func (odp *operationDataFieldParser) parse(dataField []byte, sender, receiver []byte, ignoreRelayed bool) *ResponseParseData {
+	responseParse := &ResponseParseData{
+		Operation: operationTransfer,
+	}
+
+	function, args, err := odp.argsParser.ParseData(string(dataField))
+	if err != nil {
+		return responseParse
+	}
+
+	switch function {
+	case core.BuiltInFunctionESDTTransfer:
+		return odp.parseESDTTransfer(args, sender, receiver)
+	case core.BuiltInFunctionESDTNFTTransfer:
+		return odp.parseESDTNFTTransfer(args, sender, receiver)
+	case core.BuiltInFunctionMultiESDTNFTTransfer:
+		return odp.parseMultiESDTNFTTransfer(args, sender, receiver)
+	case core.BuiltInFunctionESDTLocalBurn, core.BuiltInFunctionESDTLocalMint:
+		return parseQuantityOperationESDT(args, function)
+	case core.BuiltInFunctionESDTWipe, core.BuiltInFunctionESDTFreeze, core.BuiltInFunctionESDTUnFreeze:
+		return parseBlockingOperationESDT(args, function)
+	case core.BuiltInFunctionESDTNFTCreate, core.BuiltInFunctionESDTNFTBurn, core.BuiltInFunctionESDTNFTAddQuantity:
+		return parseQuantityOperationNFT(args, function)
+	case core.RelayedTransaction, core.RelayedTransactionV2:
+		if ignoreRelayed {
+			return &ResponseParseData{
+				IsRelayed: true,
+			}
+		}
+		return odp.parseRelayed(function, args, receiver)
+	}
+
+	if function != "" && core.IsSmartContractAddress(receiver) {
+		responseParse.Function = function
+	}
+
+	return responseParse
+}
+
+func (odp *operationDataFieldParser) parseRelayed(function string, args [][]byte, receiver []byte) *ResponseParseData {
+	if len(args) == 0 {
+		return &ResponseParseData{
+			IsRelayed: true,
+		}
+	}
+
+	tx, ok := extractInnerTx(function, args, receiver)
+	if !ok {
+		return &ResponseParseData{
+			IsRelayed: true,
+		}
+	}
+
+	res := odp.parse(tx.Data, tx.SndAddr, tx.RcvAddr, true)
+	if res.IsRelayed {
+		return &ResponseParseData{
+			IsRelayed: true,
+		}
+	}
+
+	receivers := [][]byte{tx.RcvAddr}
+	receiversShardID := []uint32{odp.shardCoordinator.ComputeId(tx.RcvAddr)}
+	if res.Operation == core.BuiltInFunctionMultiESDTNFTTransfer || res.Operation == core.BuiltInFunctionESDTNFTTransfer {
+		receivers = res.Receivers
+		receiversShardID = res.ReceiversShardID
+	}
+
+	return &ResponseParseData{
+		Operation:        res.Operation,
+		Function:         res.Function,
+		ESDTValues:       res.ESDTValues,
+		Tokens:           res.Tokens,
+		Receivers:        receivers,
+		ReceiversShardID: receiversShardID,
+		IsRelayed:        true,
+	}
+}
+
+func extractInnerTx(function string, args [][]byte, receiver []byte) (*transaction.Transaction, bool) {
+	tx := &transaction.Transaction{}
+
+	if function == core.RelayedTransaction {
+		err := json.Unmarshal(args[0], &tx)
+
+		return tx, err == nil
```

Comment:
I will let as it is because I don;t need to return error here

Reply:
Right, this func returns a bool instead of an error.
Could be useful at least to add a log there in case of an error to know why the parsing failed? (since we do not return the error)

Reply:
In this case no, because if a user does something wrong when it serializes the inner tx the relayed transaction will fail and it is enough to say the operation is relayed.